File: tfs-ticketservice-ref/web/scripts/TicketService.ts
import * as Q from "q";

export class Ticket {
    public Id: number;
    public Description: string;
    public Title: string;
    public ServiceType: string;
    public Url: string;

    constructor(id?: number) {
        if (id) {
            this.Id = id;
        }
    }
}

export interface ITicketService {

    readonly ServiceUrl: string;
    retrieveTicket(id: number): Q.IPromise<Ticket>;
    updateTicket(t: Ticket): Q.IPromise<Ticket>;
}

export class TicketService implements ITicketService {

    private _serviceUrl;

    public get ServiceUrl(): string {
        return this._serviceUrl;
    }

    constructor(serviceUrl: string) {

        this._serviceUrl = serviceUrl;
    }

    public retrieveTicket(id: number): Q.Promise<Ticket> {

        var url = this._serviceUrl + "tickets/" + id;

        return Q($.ajax({
            url: url,
            xhrFields: { withCredentials: true },
            method: "GET",
            dataType: "json"
        })).catch((e) => {
            console.error("GET to " + url + " returned an error: " + JSON.stringify(e));
            throw e;
        });
    }

    public updateTicket(t: Ticket): Q.Promise<Ticket> {

        if (t.Id == undefined || t.Id == null) {
           return Q.reject<Ticket>(new Error("task.id cannot be null or undefined"));
        }

        var url = this._serviceUrl + "tickets/" + t.Id;

        return Q($.ajax({
            url: url,
            xhrFields: { withCredentials: true },
            method: "PUT",
            data: t, dataType: "json"
        })).catch((e) => {
            console.error("PUT to " + url + " with data " + JSON.stringify(t) + " returned an error: " + JSON.stringify(e));
            throw e;
        });
    }
}

File: tfs-ticketservice-ref/api/Models/TicketService.cs
using Newtonsoft.Json.Linq;
using System;
using System.Net.Http;
using System.Text;

namespace TfsTicketServiceRef.Models
{
    //Fake implementation fo the sake of this exercise
    public class TicketService : ITicketService
    {

        public Ticket RetrieveTicket(int id)
        {
            return new Ticket() {
                Id = id,
                Description = $"Description for the ticket #{id}",
                ServiceType = "Change Request",
                Title = $"Title for the ticket #{id}",
                Url = $"https://some-url.com/view/tickets/{id}"
            };
        }

        public Ticket UpdateTicket(Ticket update)
        {
            Ticket ticket = new Ticket();
            ticket.Id = update.Id;
            ticket.Title = update.Title ?? $"Title for the ticket #{update.Id}";
            ticket.Description = update.Description ?? $"Description for the ticket #{update.Id}";
            ticket.ServiceType = update.ServiceType ?? "Change Request";
            ticket.Url = $"https://some-url.com/view/tickets/{update.Url}";

            return ticket;
        }   
    }
}